File: app/src/main/java/com/fallgamlet/dnestrcinema/utils/DateTimeUtils.java
package com.fallgamlet.dnestrcinema.utils;

import androidx.annotation.NonNull;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;

/**
 * Created by fallgamlet on 05.07.16.
 */
public class DateTimeUtils {
    //region Static fields
    private static SimpleDateFormat mDateFormatWithoutTimezone;
    private static SimpleDateFormat mDateFormatISO;
    private static SimpleDateFormat mDateFormatDotWithoutTime;
    private static SimpleDateFormat mDateFormatTimeOnly;
    private static SimpleDateFormat mDateFormatDot;
    private static SimpleDateFormat mDateFormatNamed;
    private static SimpleDateFormat mDateFormatDayOnlyFullNamed;

    public static final long SECOND = 1000;
    public static final long MINUTE = 60*SECOND;
    public static final long HOUR = 60*MINUTE;
    public static final long DAY = 24*HOUR;
    //endregion

    //region Methods
    private static SimpleDateFormat getDateFormatISO() {
        if (mDateFormatISO == null) { mDateFormatISO = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ", Locale.getDefault()); }
        return mDateFormatISO;
    }
    public static String getISO8601Date(Date date) {
        if (date == null) { return null; }
        SimpleDateFormat dateFormat = getDateFormatISO();
        String s =  dateFormat.format(date);
        if (dateFormat.toPattern().contains("Z")) {
            return new StringBuilder(s).insert(s.length()-2, ':').toString();
        } else {
            return s;
        }
    }
    public static String getISODate(Date date) {
        if (date == null) { return null; }
        SimpleDateFormat dateFormat = getDateFormatISO();
        return dateFormat.format(date);
    }
    public static Date iso8601ToDate(String datetimeStr) {
        if (datetimeStr == null) { return null; }
        SimpleDateFormat dateFormat = getDateFormatISO();
        Date date = null;
        try {
            date = dateFormat.parse(datetimeStr);
        } catch (Exception ignore) {}
        return date;
    }

    private static SimpleDateFormat getDateFormatWithoutTimezone() {
        if (mDateFormatWithoutTimezone == null) { mDateFormatWithoutTimezone = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.getDefault()); }
        return mDateFormatWithoutTimezone;
    }
    public static String getISODateWithoutTimezone(Date date) {
        if (date == null) { return null; }
        SimpleDateFormat dateFormat = getDateFormatWithoutTimezone();
        return dateFormat.format(date);
    }
    public static Date isoWithoutTimezoneToDate(String datetimeStr) {
        if (datetimeStr == null) { return null; }
        SimpleDateFormat dateFormat = getDateFormatWithoutTimezone();
        Date date = null;
        try {
            date = dateFormat.parse(datetimeStr);
        } catch (Exception ignore) {}
        return date;
    }

    private static SimpleDateFormat getDateFormatDotWithoutTime() {
        if (mDateFormatDotWithoutTime == null) { mDateFormatDotWithoutTime = new SimpleDateFormat("dd.MM.yyyy", Locale.getDefault()); }
        return mDateFormatDotWithoutTime;
    }
    public static String getDateDotWithoutTime(Date date) {
        if (date == null) { return null; }
        SimpleDateFormat dateFormat = getDateFormatDotWithoutTime();
        return dateFormat.format(date);
    }

    private static SimpleDateFormat getDateFormatTimeOnly() {
        if (mDateFormatTimeOnly == null) { mDateFormatTimeOnly = new SimpleDateFormat("HH:mm", Locale.getDefault()); }
        return mDateFormatTimeOnly;
    }
    public static String getTimeOnly(Date date) {
        if (date == null) { return null; }
        return getDateFormatTimeOnly().format(date);
    }

    private static SimpleDateFormat getDateFormatDot() {
        if (mDateFormatDot == null) { mDateFormatDot = new SimpleDateFormat("dd.MM.yyyy HH:mm:ss", Locale.getDefault()); }
        return mDateFormatDot;
    }
    public static String getDateDot(Date date) {
        if (date == null) { return null; }
        SimpleDateFormat dateFormat = getDateFormatDot();
        return dateFormat.format(date);
    }

    private static SimpleDateFormat getDateFormatNamed() {
        if (mDateFormatNamed == null) { mDateFormatNamed = new SimpleDateFormat("d MMMM yyyy", Locale.getDefault()); }
        return mDateFormatNamed;
    }
    public static String getDateNamed(Date date) {
        if (date == null) { return null; }
        SimpleDateFormat dateFormat = getDateFormatNamed();
        return dateFormat.format(date);
    }

    private static SimpleDateFormat getDateFormatDayOnlyFullNamed() {
        if (mDateFormatDayOnlyFullNamed == null) { mDateFormatDayOnlyFullNamed = new SimpleDateFormat("E, dd MMMM", Locale.getDefault()); }
        return mDateFormatDayOnlyFullNamed;
    }
    public static String getDateDayOnlyFullNamed(Date date) {
        if (date == null) { return null; }
        SimpleDateFormat dateFormat = getDateFormatDayOnlyFullNamed();
        return dateFormat.format(date);
    }

    public static Date getDateWithoutTime(Calendar calendar, Date date) {
        if (calendar == null || date == null) { return null; }
        calendar.setTime(date);
        calendar.set(Calendar.HOUR_OF_DAY, 0);
        calendar.set(Calendar.MINUTE, 0);
        calendar.set(Calendar.SECOND, 0);
        calendar.set(Calendar.MILLISECOND, 0);
        return calendar.getTime();
    }
    public static Date getDateWithoutTime(Date date) {
        if (date == null) { return null; }
        Calendar calendar = Calendar.getInstance();
        return getDateWithoutTime(calendar, date);
    }

    public static Date addDays(Date date, int days) {
        if (date == null) { return null; }
        Calendar calendar = Calendar.getInstance();
        calendar.setTime(date);
        calendar.add(Calendar.DAY_OF_YEAR, days);
        return calendar.getTime();
    }
    public static Date addHours(Date date, int hours) {
        if (date == null) { return null; }
        Calendar calendar = Calendar.getInstance();
        calendar.setTime(date);
        calendar.add(Calendar.HOUR_OF_DAY, hours);
        return calendar.getTime();
    }
    public static Date addMinutes(Date date, int minutes) {
        if (date == null) { return null; }
        Calendar calendar = Calendar.getInstance();
        calendar.setTime(date);
        calendar.add(Calendar.MINUTE, minutes);
        return calendar.getTime();
    }
    public static Date addMilliseconds(Date date, int milliseconds) {
        if (date == null) { return null; }
        long time = date.getTime() + milliseconds;
        return new Date(time);
    }

    public static long getDuration(@NonNull Date start, @NonNull Date end) {
        return end.getTime() - start.getTime();

    }

    //region Month names singletons
    private static String[] monthNamesByRussian = null;
    public static String[] getMonthNamesByRussian() {
        if (monthNamesByRussian == null) {
            monthNamesByRussian = new String[] {"январь","февраль","март","апрель","май","июнь","июль","август","сентябрь","октябрь","ноябрь","декабрь"};
        }
        return monthNamesByRussian;
    }

    private static String[] monthNamesByEnglish = null;
    public static String[] getMonthNamesByEnglish() {
        if (monthNamesByEnglish == null) {
            monthNamesByEnglish = new String[] {"january", "february", "march", "april", "may", "june", "july", "august", "september", "october", "november", "december"};
        }
        return monthNamesByEnglish;
    }
    //endregion

    public static String getMonthName(int month) {
        return getMonthName(month, Locale.getDefault());
    }
    public static String getMonthName(int month, Locale locale) {
        if (locale == null) { return null; }
        String name = null;
        String[] months = null;
        String lang = locale.getLanguage();
        if (lang.equals("ru")) { months = getMonthNamesByRussian(); }
        else { months = getMonthNamesByEnglish(); }

        if (0<=month && month<months.length) { name = months[month]; }
        return name;
    }

    public static String getNamedMonthYear(Date date, Locale locale) {
        Calendar calendar = Calendar.getInstance();
        calendar.setTime(date);
        return getNamedMonthYear(calendar, locale);
    }
    public static String getNamedMonthYear(Calendar date, Locale locale) {
        if (date == null) { return null; }
        int month = date.get(Calendar.MONTH);
        int year = date.get(Calendar.YEAR);
        String monthName = getMonthName(month, locale);
        return monthName+" "+year;
    }
    //endregion
}

File: app/src/main/java/com/fallgamlet/dnestrcinema/utils/DateTimeUtils.kt
package com.fallgamlet.dnestrcinema.utils

import android.annotation.SuppressLint
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale


@SuppressLint("ConstantLocale")
object DateTimeUtils {

    private val dateFormatWithoutTimezone: SimpleDateFormat by lazy { SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", locale) }
    private val dateFormatISO: SimpleDateFormat by lazy { SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ", locale) }
    private val dateFormatDotWithoutTime: SimpleDateFormat by lazy { SimpleDateFormat("dd.MM.yyyy", locale) }
    private val dateFormatTimeOnly: SimpleDateFormat by lazy { SimpleDateFormat("HH:mm", locale) }
    private val dateFormatDot: SimpleDateFormat by lazy { SimpleDateFormat("dd.MM.yyyy HH:mm:ss", locale) }
    private val dateFormatNamed: SimpleDateFormat by lazy { SimpleDateFormat("d MMMM yyyy", locale) }
    private val dateFormatDayOnlyFullNamed: SimpleDateFormat by lazy { SimpleDateFormat("E, dd MMMM", locale) }
    private val locale = Locale.getDefault()

    private val monthNamesByRussian: Array<String> by lazy { arrayOf("январь", "февраль", "март", "апрель", "май", "июнь", "июль", "август", "сентябрь", "октябрь", "ноябрь", "декабрь") }
    private val monthNamesByEnglish: Array<String> by lazy { arrayOf("january", "february", "march", "april", "may", "june", "july", "august", "september", "october", "november", "december") }

    val SECOND: Long = 1000
    val MINUTE = 60 * SECOND
    val HOUR = 60 * MINUTE
    val DAY = 24 * HOUR

    fun getISO8601Date(date: Date?): String? {
        date ?: return null
        val dateFormat = dateFormatISO
        val s = dateFormat.format(date)
        return if (dateFormat.toPattern().contains("Z")) {
            StringBuilder(s).insert(s.length - 2, ':').toString()
        } else {
            s
        }
    }

    fun getISODate(date: Date?): String? {
        date ?: return null
        return dateFormatISO.format(date)
    }

    fun iso8601ToDate(datetimeStr: String?): Date? {
        datetimeStr ?: return null
        var date: Date? = null
        try {
            date = dateFormatISO.parse(datetimeStr)
        } catch (ignore: Exception) {
        }

        return date
    }

    fun getISODateWithoutTimezone(date: Date?): String? {
        date ?: return null
        return dateFormatWithoutTimezone.format(date)
    }

    fun isoWithoutTimezoneToDate(datetimeStr: String?): Date? {
        datetimeStr ?: return null
        return try {
            dateFormatWithoutTimezone.parse(datetimeStr)
        } catch (ignore: Exception) {
            null
        }
    }

    fun getDateDotWithoutTime(date: Date?): String? {
        date ?: return null
        return dateFormatDotWithoutTime.format(date)
    }

    fun getTimeOnly(date: Date?): String? {
        return if (date == null) {
            null
        } else dateFormatTimeOnly.format(date)
    }

    fun getDateDot(date: Date?): String? {
        date ?: return null
        return dateFormatDot.format(date)
    }

    fun getDateNamed(date: Date?): String? {
        date ?: return null
        return dateFormatNamed.format(date)
    }

    fun getDateDayOnlyFullNamed(date: Date?): String? {
        date ?: return null
        return dateFormatDayOnlyFullNamed.format(date)
    }

    fun getDateWithoutTime(calendar: Calendar?, date: Date?): Date? {
        calendar ?: return null
        date ?: return null
        calendar.time = date
        calendar.set(Calendar.HOUR_OF_DAY, 0)
        calendar.set(Calendar.MINUTE, 0)
        calendar.set(Calendar.SECOND, 0)
        calendar.set(Calendar.MILLISECOND, 0)
        return calendar.time
    }

    fun getDateWithoutTime(date: Date?): Date? {
        date ?: return null
        val calendar = Calendar.getInstance()
        return getDateWithoutTime(calendar, date)
    }

    fun addDays(date: Date?, days: Int): Date? {
        date ?: return null
        val calendar = Calendar.getInstance()
        calendar.time = date
        calendar.add(Calendar.DAY_OF_YEAR, days)
        return calendar.time
    }

    fun addHours(date: Date?, hours: Int): Date? {
        date ?: return null
        val calendar = Calendar.getInstance()
        calendar.time = date
        calendar.add(Calendar.HOUR_OF_DAY, hours)
        return calendar.time
    }

    fun addMinutes(date: Date?, minutes: Int): Date? {
        date ?: return null
        val calendar = Calendar.getInstance()
        calendar.time = date
        calendar.add(Calendar.MINUTE, minutes)
        return calendar.time
    }

    fun addMilliseconds(date: Date?, milliseconds: Int): Date? {
        date ?: return null
        val time = date.time + milliseconds
        return Date(time)
    }

    fun getDuration(start: Date, end: Date): Long {
        return end.time - start.time
    }



    @JvmOverloads
    fun getMonthName(month: Int, locale: Locale? = Locale.getDefault()): String {
        locale ?: return ""

        val months = when (locale.language) {
            "ru" -> monthNamesByRussian
            else -> monthNamesByEnglish
        }

        return when (0 <= month && month < months.size) {
            true -> months[month]
            else -> ""
        }
    }

    fun getNamedMonthYear(date: Date, locale: Locale): String {
        val calendar = Calendar.getInstance()
        calendar.time = date
        return getNamedMonthYear(calendar, locale)
    }

    fun getNamedMonthYear(date: Calendar?, locale: Locale): String {
        date ?: return ""
        val month = date.get(Calendar.MONTH)
        val year = date.get(Calendar.YEAR)
        val monthName = getMonthName(month, locale)
        return "$monthName $year"
    }

}
